sdks/java/io/mqtt/src/main/java/org/apache/beam/sdk/io/mqtt/MqttIO.java

```diff
@@ -0,0 +1,571 @@
+      }
+      return currentTimestamp;
+    }
+
+    @Override
+    public UnboundedMqttSource getCurrentSource() {
+      return source;
+    }
+
+  }
+
+  /**
+   * A {@link PTransform} to write and send a message to a MQTT server.
+   */
+  @AutoValue
+  public abstract static class Write extends PTransform<PCollection<byte[]>, PDone> {
+
+    @Nullable abstract ConnectionConfiguration connectionConfiguration();
+    @Nullable abstract int qos();
+    @Nullable abstract boolean retained();
+
+    abstract Builder toBuilder();
+
+    @AutoValue.Builder
+    abstract static class Builder {
+      abstract Builder setConnectionConfiguration(ConnectionConfiguration configuration);
+      abstract Builder setQos(int qos);
+      abstract Builder setRetained(boolean retained);
+      abstract Write build();
+    }
+
+    /**
+     * Define MQTT connection configuration used to connect to the MQTT broker.
+     */
+    public Write withConnectionConfiguration(ConnectionConfiguration configuration) {
+      return toBuilder().setConnectionConfiguration(configuration).build();
+    }
+
+    /**
+     * Define the MQTT message quality of service.
+     *
+     * <ul>
+     * <li>Quality of Service 0 - indicates that a message should
+     * be delivered at most once (zero or one times). The message will not be persisted to disk,
+     * and will not be acknowledged across the network. This QoS is the fastest,
+     * but should only be used for messages which are not valuable - note that
+     * if the server cannot process the message (for example, there
+     * is an authorization problem), then an
+     * {@link MqttCallback#deliveryComplete(IMqttDeliveryToken)} won't be called.
+     * Also known as "fire and forget".</li>
+     *
+     * <li>Quality of Service 1 - indicates that a message should
+     * be delivered at least once (one or more times).  The message can only be delivered safely if
+     * it can be persisted on the broker.
+     * If the broker can't persist the message, the message will not be
+     * delivered in the event of a client failure.
+     * The message will be acknowledged across the network.
+     * This is the default QoS.</li>
+     *
+     * <li>Quality of Service 2 - indicates that a message should
+     * be delivered once. The message will be persisted to disk on the broker, and will
+     * be subject to a two-phase acknowledgement across the network.
+     * The message can only be delivered safely if
+     * it can be persisted on the broker.
+     * If a persistence mechanism is not specified on the broker, the message will not be
+     * delivered in the event of a client failure.</li>
+     * </ul>
+     *
+     * <p>If persistence is not configured, QoS 1 and 2 messages will still be delivered
+     * in the event of a network or server problem as the client will hold state in memory.
+     * If the MQTT client is shutdown or fails and persistence is not configured then
```

Comment:
I'd suggest to rephrase this in terms of consequences for someone who uses this transform in a pipeline. The pipeline writer has no control over the MQTT client - it runs inside a bundle that runs inside a VM or container or somewhere else, controlled by the runner and subject to arbitrary failures that the runner transparently handles and retries.

So basically, for a pipeline writer, this means that messages can be duplicated (in case the writer bundle is retried multiple times), but can not be lost: at-least-once delivery.

By the way, this means specifying QoS 2 is meaningless, since we can not provide exactly-once guarantees anyway. In order to get those guarantees, MQTT protocol would need to provide deduplication-by-message-id - does it have a feature like that? If yes, how can the pipeline writer control message ids? (automatically generated ones are not sufficient - see my latest email to beam-dev@ about the Checkpoint transform)

This might be too complicated to address in the current PR, so I'd suggest to just document that this class provides best-effort delivery at QoS 0 and at-least-once delivery at QoS 1, and address exactly-once delivery later.